{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ebc19951-a31c-4292-aa65-8eb6f63a765f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# AlloBench Dataset Creation Pipeline\n",
    "\n",
    "TODO: Fix Active Site annotation with respect to PDB and not Uniprot\n",
    "\n",
    "Merge on PDB ID and Uniprot ID to fix obsolete UniProt IDs\n",
    "\n",
    "Add Protein Name Organism from PDB and UniProt\n",
    "\n",
    "Check PDB files not available as .pdb and download mmCIF and convert to PDB if missing residues model and save structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a4d8b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_asd_download = 'ASD_Release_202306_XF'\n",
    "asd_csv_output = \"ASD_Updated.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "553f3a34",
   "metadata": {},
   "source": [
    "## Get AlloSteric Database (ASD) Data\n",
    "\n",
    "Download the ASD_Release_202306_XF.tar.gz from https://mdl.shsmu.edu.cn/ASD/module/download/download.jsp?tabIndex=1 and extract it here or provide the path to the extracted directory to parse_asd_xml()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22af2d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import tqdm\n",
    "import xmltodict\n",
    "\n",
    "\n",
    "def parse_allosteric_site(alloteric_site):\n",
    "    \"\"\"Convert the list of residues in ASD from \"Chain A:HIS25,TYR258; Chain B:VAL325\" to ['A-HIS-25', 'A-TYR-258', 'B-VAL-325']\"\"\"\n",
    "    residues = []\n",
    "    for chain_string in alloteric_site.split('; '):\n",
    "        chain_name, residue_string = chain_string.split(':')\n",
    "        chain_id = chain_name[-1]\n",
    "        for residue in residue_string.split(','):\n",
    "            res_name, res_id = residue[:3], residue[3:]\n",
    "            residues.append(f'{chain_id}-{res_name}-{res_id}')\n",
    "    return residues\n",
    "\n",
    "\n",
    "def parse_asd_xml(asd_xml_dir):\n",
    "    \"\"\"parse a directory containing XML files from the ASD database to create a pandas DataFrame\"\"\"\n",
    "    data = []\n",
    "    for xml_file in tqdm.tqdm(glob.glob(os.path.join(asd_xml_dir, '*.xml'))):\n",
    "        xml_string = open(xml_file, 'r').read()\n",
    "        xml_string = xml_string.replace('&#x2;', '')  # Remove invalid XML character\n",
    "        protein = xmltodict.parse(xml_string)\n",
    "    \n",
    "        target_gene = ''\n",
    "        target_id = protein['Organism_Record']['Organism_ID']\n",
    "    \n",
    "        if 'Gene' in protein['Organism_Record'] and 'Gene_Name' in protein['Organism_Record']['Gene']:\n",
    "            target_gene = protein['Organism_Record']['Gene']['Gene_Name']\n",
    "    \n",
    "        organism = protein['Organism_Record']['Organism']\n",
    "    \n",
    "        # Check if the list of allosteric sites in present in the XML file\n",
    "        if 'Allosteric_Site_List' in protein['Organism_Record']:\n",
    "            allosteric_sites = protein['Organism_Record']['Allosteric_Site_List']['Allosteric_Site']\n",
    "    \n",
    "            # Check if more than one allosteric site is present\n",
    "            if isinstance(allosteric_sites, list):\n",
    "                allosteric_sites = allosteric_sites\n",
    "            else:\n",
    "                allosteric_sites = [allosteric_sites]\n",
    "        else:\n",
    "            allosteric_sites = []\n",
    "    \n",
    "        for site in allosteric_sites:\n",
    "            if site:\n",
    "                pdb_uniprot = site['PDB_UniProt_ID'] if 'PDB_UniProt_ID' in site else ''\n",
    "                allosteric_pdb = site['Allosteric_PDB']\n",
    "                modulator_serial = site['Modulator_ASD_ID']\n",
    "                modulator_alias = site['Modulator_Alias']\n",
    "                modulator_chain = site['Modulator_Chain']\n",
    "                modulator_class = site['Modulator_Class'] if 'Modulator_Class' in site else ''\n",
    "                modulator_feature = site['Modulator_Feature'] if 'Modulator_Feature' in site else ''\n",
    "                modulator_name = site['Modulator_Name'] if 'Modulator_Name' in site else ''\n",
    "                modulator_resi = site['Modulator_Residue'] if 'Modulator_Residue' in site else ''\n",
    "                function = site['Function'] if 'Function' in site else ''\n",
    "                position = site['Position'] if 'Position' in site else ''\n",
    "                pubmed_id = site['PubMed_ID'] if 'PubMed_ID' in site else ''\n",
    "                ref_title = site['PubMed_Title'] if 'PubMed_Title' in site else ''\n",
    "                site_overlap = site['Site_Overlap'] if 'Site_Overlap' in site else ''\n",
    "                allosteric_site_residue = parse_allosteric_site(site['Allosteric_Site_Residue']) if 'Allosteric_Site_Residue' in site else []\n",
    "    \n",
    "                data.append([\n",
    "                    target_id, target_gene, organism, pdb_uniprot, allosteric_pdb,\n",
    "                    modulator_serial, modulator_alias, modulator_chain,\n",
    "                    modulator_class, modulator_feature, modulator_name,\n",
    "                    modulator_resi, function, position, pubmed_id, ref_title,\n",
    "                    site_overlap, allosteric_site_residue])\n",
    "    \n",
    "    return pd.DataFrame(data, columns=[\n",
    "        'target_id', 'target_gene', 'organism', 'pdb_uniprot', 'allosteric_pdb',\n",
    "        'modulator_serial', 'modulator_alias', 'modulator_chain', 'modulator_class',\n",
    "        'modulator_feature', 'modulator_name', 'modulator_resi', 'function',\n",
    "        'position', 'pubmed_id', 'ref_title', 'site_overlap',\n",
    "        'allosteric_site_residue'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83e3055a-f540-4b7a-a89a-102ee6748b56",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2419/2419 [00:03<00:00, 772.91it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Rows:              3172\n",
      "Number of Unique PDB IDs:    2993\n",
      "Number of Unique UniProt AC: 701\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "df_asd = parse_asd_xml(path_to_asd_download)\n",
    "\n",
    "# Make the UniProt and PDB IDs uppercase to facilitate merging on these later\n",
    "df_asd['allosteric_pdb'] = df_asd['allosteric_pdb'].str.upper()\n",
    "df_asd['pdb_uniprot'] = df_asd['pdb_uniprot'].str.upper()\n",
    "\n",
    "print('Number of Rows:             ', df_asd.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_asd['allosteric_pdb'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_asd['pdb_uniprot'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3dc35a2-f36f-43e5-89fe-7d5d47bc02e4",
   "metadata": {},
   "source": [
    "## Fix Obsolete PDB Entries\n",
    "\n",
    "Some PDB IDs in the allosteric_pdb column are obsolete. The list of obsolete PDB IDs is downloaded from the Protein Data Bank. The obsolete PDB IDs are replaced by the superseded PDB IDs and structures without any superseded PDB IDs are removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc1b70b3-a462-48c3-97b1-a9381d78c8a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import gzip\n",
    "\n",
    "response = requests.get('https://files.wwpdb.org/pub/pdb/holdings/all_removed_entries.json.gz')\n",
    "\n",
    "obsolete_pdb = {}\n",
    "for pdb_id, value in json.loads(gzip.decompress(response.content)).items():\n",
    "    if 'superseded_by' in value:\n",
    "        # Select the last element in the list of superseded PDB IDs\n",
    "        obsolete_pdb[pdb_id] = value['superseded_by'] # [-1] \n",
    "    else:\n",
    "        obsolete_pdb[pdb_id] = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "64845d72-5118-4020-858f-59e476deacab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Obsolete PDB ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1HWZ</th>\n",
       "      <td>6DHD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3MVQ</th>\n",
       "      <td>6DHL</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3MW9</th>\n",
       "      <td>6DHM</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3QMU</th>\n",
       "      <td>6DHQ</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4GAR</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4GAU</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4LEG</th>\n",
       "      <td>5J94</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0     1     2\n",
       "Obsolete PDB ID                  \n",
       "1HWZ             6DHD  None  None\n",
       "3MVQ             6DHL  None  None\n",
       "3MW9             6DHM  None  None\n",
       "3QMU             6DHQ  None  None\n",
       "4GAR             4V9C  None  None\n",
       "4GAU             4V9C  None  None\n",
       "4LEG             5J94  None  None"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_obsolete_pdb = pd.DataFrame.from_dict(obsolete_pdb, orient='index')\n",
    "df_obsolete_pdb.index.name = 'Obsolete PDB ID'\n",
    "\n",
    "# Obsolete PDB IDs in ASD\n",
    "df_obsolete_pdb_asd = df_obsolete_pdb[df_obsolete_pdb.index.isin(df_asd['allosteric_pdb'])]\n",
    "\n",
    "# Replace the obsolete PDB IDs\n",
    "df_asd['allosteric_pdb'] = df_asd['allosteric_pdb'].replace(df_obsolete_pdb_asd.to_dict()[0])\n",
    "\n",
    "# Check if any of the obsolete PDB IDs had more than one updated PDB IDs\n",
    "df_obsolete_pdb_asd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a079acea-5f02-40aa-9733-c25884563a14",
   "metadata": {},
   "source": [
    "While most obsolete PDB IDs are superseded by only one PDB ID, there exists obsolete PDB IDs superseded by multiple PDB IDs. However, the ASD does not contain such structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3855d8e7-57ac-42ce-9ce0-e0b3078fbe4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Rows:              3172\n",
      "Number of Unique PDB IDs:    2991\n",
      "Number of Unique UniProt AC: 701\n"
     ]
    }
   ],
   "source": [
    "print('Number of Rows:             ', df_asd.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_asd['allosteric_pdb'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_asd['pdb_uniprot'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c17fcf-6e50-4b89-89cb-2b934f9aecd4",
   "metadata": {},
   "source": [
    "## Update UniProt IDs\n",
    "\n",
    "The active site information will be obtained from the UniProt database hence it is critical to have the correct UniProt ID for the ASD protein structures. The code to fix the obsolete UniProt IDs is in the fix_uniprot_id.py\n",
    "\n",
    "1. Get the list of UniProt IDs in the structure corresponding to each PDB ID in the ASD from the PDB using GraphQL  \n",
    "2. If the UniProt ID from ASD is present in the UniProt IDs fetched from the PDB then the UniProt ID from ASD is correct and does not need any modification.  \n",
    "3. If the UniProt ID from ASD is absent in the UniProt IDs fetched from the PDB then,  \n",
    "    a. If only 1 UniProt ID was obtained then that ID may be assgined without ambiguity  \n",
    "    b. If 0 or multiple UniProt IDs were obtained then these must be reviewed and any suitable UniProt IDs are included in the  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "216fa124",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniprot_from_pdb(pdb_ids):\n",
    "    \"\"\" Get the UniProt IDs for the chains in the structures with the PDB IDs \"\"\"\n",
    "    pdb_ids_string = '\", \"'.join(pdb_ids)\n",
    "\n",
    "    body = 'query {entries(entry_ids: [\"' + pdb_ids_string + '\"])' + \"\"\"\n",
    "      {\n",
    "        rcsb_id\n",
    "        polymer_entities {\n",
    "          uniprots {\n",
    "            rcsb_id\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \"\"\"\n",
    "    response = requests.post(url='https://data.rcsb.org/graphql', json={\"query\": body})\n",
    "    response_data = response.json()\n",
    "\n",
    "    id_mapping = []\n",
    "    for record in response_data['data']['entries']:\n",
    "        rcsb_id = record['rcsb_id']\n",
    "        uniprot_ids = set()\n",
    "        for entity in record['polymer_entities']:\n",
    "            if entity['uniprots']:\n",
    "                uniprot_ids.add(entity['uniprots'][0]['rcsb_id'])\n",
    "        id_mapping.append([rcsb_id, uniprot_ids])\n",
    "        \n",
    "    return pd.DataFrame(id_mapping, columns=['allosteric_pdb', 'new_uniprot_id'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8bd85b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_asd_subset = df_asd[['allosteric_pdb', 'pdb_uniprot']].drop_duplicates().copy()\n",
    "\n",
    "# Get the UniProt IDs of the chains in the proteins from PDB website\n",
    "df_pdb = uniprot_from_pdb(df_asd_subset['allosteric_pdb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef6ca756",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDB IDs Absent from ASD with Zero or Multiple UniProt ID fetched from the PDB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allosteric_pdb</th>\n",
       "      <th>pdb_uniprot</th>\n",
       "      <th>new_uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2426</th>\n",
       "      <td>1CKK</td>\n",
       "      <td>P62158</td>\n",
       "      <td>{P0DP33, P97756}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2424</th>\n",
       "      <td>1IQ5</td>\n",
       "      <td>P62158</td>\n",
       "      <td>{P0DP33, Q3Y416}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2423</th>\n",
       "      <td>1NWD</td>\n",
       "      <td>P62158</td>\n",
       "      <td>{P0DP33, Q07346}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>3J41</td>\n",
       "      <td>P62158</td>\n",
       "      <td>{P0DP23, Q6J8I9}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>3OBK</td>\n",
       "      <td>S8F7E9</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557</th>\n",
       "      <td>3OWI</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>3OWW</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554</th>\n",
       "      <td>3OWZ</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1566</th>\n",
       "      <td>3OXE</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>3OXJ</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1555</th>\n",
       "      <td>3OXM</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1559</th>\n",
       "      <td>3Q3Z</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>3RHW</td>\n",
       "      <td>O17793</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>3RI5</td>\n",
       "      <td>O17793</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>3RIA</td>\n",
       "      <td>O17793</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>3RIF</td>\n",
       "      <td>O17793</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>4A2U</td>\n",
       "      <td>I6Y496</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2956</th>\n",
       "      <td>4FHA</td>\n",
       "      <td>Q9I4W3</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2957</th>\n",
       "      <td>4I7W</td>\n",
       "      <td>Q9I4W3</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>4KGV</td>\n",
       "      <td>E8Y329</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>4KGX</td>\n",
       "      <td>E8Y329</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>4KH0</td>\n",
       "      <td>E8Y329</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>4KH1</td>\n",
       "      <td>E8Y329</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2623</th>\n",
       "      <td>4P86</td>\n",
       "      <td>P39765</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2941</th>\n",
       "      <td>4QSH</td>\n",
       "      <td>L8DYY9</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2940</th>\n",
       "      <td>4QSK</td>\n",
       "      <td>L8DYY9</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2914</th>\n",
       "      <td>4RQO</td>\n",
       "      <td>A0A130G9H0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>P08179</td>\n",
       "      <td>{P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2502</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>Q7MAW7</td>\n",
       "      <td>{P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2503</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>Q7MAW7</td>\n",
       "      <td>{P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4V9C</td>\n",
       "      <td>P08179</td>\n",
       "      <td>{P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2170</th>\n",
       "      <td>6FMU</td>\n",
       "      <td>G4V8J4</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2169</th>\n",
       "      <td>6FMZ</td>\n",
       "      <td>G4V8J4</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2172</th>\n",
       "      <td>6FP4</td>\n",
       "      <td>G4V8J4</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171</th>\n",
       "      <td>6FTC</td>\n",
       "      <td>G4V8J4</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>6I2P</td>\n",
       "      <td>A0R2B1</td>\n",
       "      <td>{P9WI81, P9WJA9}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>6KDY</td>\n",
       "      <td>Q43837</td>\n",
       "      <td>{O43837, P50213}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>6UI4</td>\n",
       "      <td>A0A098D3M4</td>\n",
       "      <td>{I1RE19, I1RCT2}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2565</th>\n",
       "      <td>6VFN</td>\n",
       "      <td>A0A0G3E2X5</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2832</th>\n",
       "      <td>7LD3</td>\n",
       "      <td>P30542</td>\n",
       "      <td>{P08173, P62873, P59768, P04899}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>7O83</td>\n",
       "      <td>P01116</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     allosteric_pdb pdb_uniprot  \\\n",
       "2426           1CKK      P62158   \n",
       "2424           1IQ5      P62158   \n",
       "2423           1NWD      P62158   \n",
       "2425           3J41      P62158   \n",
       "1055           3OBK      S8F7E9   \n",
       "1557           3OWI               \n",
       "1563           3OWW               \n",
       "1554           3OWZ               \n",
       "1566           3OXE               \n",
       "1562           3OXJ               \n",
       "1555           3OXM               \n",
       "1559           3Q3Z               \n",
       "723            3RHW      O17793   \n",
       "722            3RI5      O17793   \n",
       "724            3RIA      O17793   \n",
       "721            3RIF      O17793   \n",
       "1131           4A2U      I6Y496   \n",
       "2956           4FHA      Q9I4W3   \n",
       "2957           4I7W      Q9I4W3   \n",
       "215            4KGV      E8Y329   \n",
       "214            4KGX      E8Y329   \n",
       "212            4KH0      E8Y329   \n",
       "213            4KH1      E8Y329   \n",
       "2623           4P86      P39765   \n",
       "2941           4QSH      L8DYY9   \n",
       "2940           4QSK      L8DYY9   \n",
       "2914           4RQO  A0A130G9H0   \n",
       "3              4V9C      P08179   \n",
       "2502           4V9C      Q7MAW7   \n",
       "2503           4V9C      Q7MAW7   \n",
       "2              4V9C      P08179   \n",
       "2170           6FMU      G4V8J4   \n",
       "2169           6FMZ      G4V8J4   \n",
       "2172           6FP4      G4V8J4   \n",
       "2171           6FTC      G4V8J4   \n",
       "702            6I2P      A0R2B1   \n",
       "110            6KDY      Q43837   \n",
       "1987           6UI4  A0A098D3M4   \n",
       "2565           6VFN  A0A0G3E2X5   \n",
       "2832           7LD3      P30542   \n",
       "683            7O83      P01116   \n",
       "\n",
       "                                         new_uniprot_id  \n",
       "2426                                   {P0DP33, P97756}  \n",
       "2424                                   {P0DP33, Q3Y416}  \n",
       "2423                                   {P0DP33, Q07346}  \n",
       "2425                                   {P0DP23, Q6J8I9}  \n",
       "1055                                                 {}  \n",
       "1557                                                 {}  \n",
       "1563                                                 {}  \n",
       "1554                                                 {}  \n",
       "1566                                                 {}  \n",
       "1562                                                 {}  \n",
       "1555                                                 {}  \n",
       "1559                                                 {}  \n",
       "723                                                  {}  \n",
       "722                                                  {}  \n",
       "724                                                  {}  \n",
       "721                                                  {}  \n",
       "1131                                                 {}  \n",
       "2956                                                 {}  \n",
       "2957                                                 {}  \n",
       "215                                                  {}  \n",
       "214                                                  {}  \n",
       "212                                                  {}  \n",
       "213                                                  {}  \n",
       "2623                                                 {}  \n",
       "2941                                                 {}  \n",
       "2940                                                 {}  \n",
       "2914                                                 {}  \n",
       "3     {P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...  \n",
       "2502  {P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...  \n",
       "2503  {P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...  \n",
       "2     {P0A7R5, P0A7R1, P0ADY7, P0A7W1, P60624, P0AA1...  \n",
       "2170                                                 {}  \n",
       "2169                                                 {}  \n",
       "2172                                                 {}  \n",
       "2171                                                 {}  \n",
       "702                                    {P9WI81, P9WJA9}  \n",
       "110                                    {O43837, P50213}  \n",
       "1987                                   {I1RE19, I1RCT2}  \n",
       "2565                                                 {}  \n",
       "2832                   {P08173, P62873, P59768, P04899}  \n",
       "683                                                  {}  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of UniProt IDs obtained from PDB website for each PDB ID\n",
    "df_pdb['num_uniprot'] = df_pdb['new_uniprot_id'].apply(len)\n",
    "df_merged = df_asd_subset.merge(df_pdb, how='left')\n",
    "\n",
    "# Check if the UniProt ID in ASD is present in the UniProt IDs obtained from PDB website\n",
    "df_merged['present'] = df_merged.apply(lambda x: x['pdb_uniprot'] in x['new_uniprot_id'], axis=1)\n",
    "df_present = df_merged.loc[df_merged['present'], ['allosteric_pdb', 'pdb_uniprot']]\n",
    "df_present['new_uniprot_id'] = df_present['pdb_uniprot']\n",
    "\n",
    "# UniProt IDs absent in the ASD but has only one UniProt ID downloaded from PDB\n",
    "df_absent_single = df_merged[~df_merged['present'] & (df_merged['num_uniprot'] == 1)].copy()\n",
    "\n",
    "# Place the only one UniProt ID from the set in Uniprot ID column to pdb_uniprot column\n",
    "if not df_absent_single.empty:\n",
    "    df_absent_single['new_uniprot_id'] = df_absent_single.apply(lambda x: next(iter(x['new_uniprot_id'])), axis=1)\n",
    "df_absent_single = df_absent_single[['allosteric_pdb', 'pdb_uniprot', 'new_uniprot_id']]\n",
    "\n",
    "# Combine the UniProt ID that are present and those returned only one UniProt ID\n",
    "df_asd_pdb_uniprot = pd.concat([df_present, df_absent_single])\n",
    "\n",
    "df_absent_multiple = df_merged[~df_merged['present'] & (df_merged['num_uniprot'] != 1)]\n",
    "print(\"PDB IDs Absent from ASD with Zero or Multiple UniProt ID fetched from the PDB\")\n",
    "df_absent_multiple[['allosteric_pdb', 'pdb_uniprot', 'new_uniprot_id']].sort_values('allosteric_pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1594ef17-8cff-4ee7-98d4-bb424f90f0da",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Reinsert PDB ID and UniProt AC combination from the manual curation of the table above\n",
    "\n",
    "The following PDB IDs from ASD were dropped: 3OWZ, 3OXM, 3OWI, 3Q3Z, 3OXJ, 3OWW, 3OXE because these are structures of glycine riboswitch, an RNA element and do not have an associated UniProt IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4c35be42-4fce-4f9c-8005-f5e4531381b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_manual_curation = pd.DataFrame(data=[\n",
    "        ['1CKK', 'P62158', 'P0DP33'],\n",
    "        ['1IQ5', 'P62158', 'P0DP33'],\n",
    "        ['1NWD', 'P62158', 'P0DP33'],\n",
    "        ['3J41', 'P62158', 'P0DP23'],\n",
    "        ['3OBK', 'S8F7E9', 'S8F7E9'],\n",
    "        ['3RHW', 'O17793', 'G5EBR3'],\n",
    "        ['3RI5', 'O17793', 'G5EBR3'],\n",
    "        ['3RIA', 'O17793', 'G5EBR3'],\n",
    "        ['3RIF', 'O17793', 'G5EBR3'],\n",
    "        ['4A2U', 'I6Y496', 'P9WP65'],\n",
    "        ['4P86', 'P39765', 'P39765'],\n",
    "        ['6KDY', 'Q43837', 'O43837'],\n",
    "        ['6UI4', 'A0A098D3M4', 'I1RCT2'],\n",
    "        ['7LD3', 'P30542', 'P30542'],\n",
    "        ['7O83', 'P01116', 'P01116']\n",
    "    ],\n",
    "    columns=['allosteric_pdb', 'pdb_uniprot', 'new_uniprot_id'])\n",
    "\n",
    "df_asd_pdb_uniprot = pd.concat([df_asd_pdb_uniprot, df_manual_curation])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "766d5ee3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Rows:              3172\n",
      "Number of Unique PDB IDs:    2991\n",
      "Number of Unique UniProt AC: 690\n"
     ]
    }
   ],
   "source": [
    "df_asd_updated =  df_asd.merge(df_asd_pdb_uniprot.drop_duplicates(), how='left')\n",
    "df_asd_updated['pdb_uniprot'] = df_asd_updated['new_uniprot_id']\n",
    "df_asd_updated.drop(columns='new_uniprot_id', inplace=True)\n",
    "\n",
    "print('Number of Rows:             ', df_asd_updated.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_asd_updated['allosteric_pdb'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_asd_updated['pdb_uniprot'].nunique())\n",
    "\n",
    "df_asd_updated.to_csv(asd_csv_output, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30f9018d-1881-4758-87e0-ace7b6429b94",
   "metadata": {},
   "source": [
    "## Merge with Additional Information from the PDB\n",
    "\n",
    "Obtain the data from PDB using its GraphQL API. Mapping between PDB chain ids and UniProt IDs. The oligomeric state, and stoichiometry of the proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09eaa8e9-4c6d-446a-9807-1ebfdd8e52d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chain_uniprot(array):\n",
    "    output = set()\n",
    "    if isinstance(array, str):\n",
    "        array = eval(array)\n",
    "    for chain in array:\n",
    "        for uniprot_id in chain[1]:\n",
    "            output.add(uniprot_id)\n",
    "    return sorted(output)\n",
    "\n",
    "\n",
    "df_pdb = pdb_utils.get_pdb_data(df_asd_updated['allosteric_pdb'].unique())\n",
    "\n",
    "df_pdb['uniprot_in_pdb'] = df_pdb['chain_uniprot_mapping'].apply(get_chain_uniprot)\n",
    "\n",
    "df_pdb.rename(columns={'pdb_id': 'allosteric_pdb'}, inplace=True)\n",
    "df_asd_pdb_merged = df_asd_updated.merge(df_pdb)\n",
    "\n",
    "print('Number of Rows:             ', df_asd_pdb_merged.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_asd_pdb_merged['allosteric_pdb'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_asd_pdb_merged['pdb_uniprot'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "888658af-7f7e-4644-9511-5ae1372e216f",
   "metadata": {},
   "source": [
    "## Download PDB files\n",
    "\n",
    "The large structures with PDB IDs 4V9C, 3BCR, 8DD3, 7AIA, 6VVH, 7APH, 7OFU are not available in .pdb format.\n",
    "\n",
    "However, these are available in PDBx/mmCIF format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b733ba13-985d-40f5-b65c-c7f5cc1070a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_utils.download_structures(\n",
    "    pdb_ids=df_asd_pdb_merged['allosteric_pdb'].unique(),\n",
    "    download_dir='../data/pdb_downloaded/'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac4b170a-b152-455c-bacb-b3725c6c3bc5",
   "metadata": {},
   "source": [
    "## Find PDB structures with missing Residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d949444-a3fe-40d1-929a-b6adc6ace060",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from tqdm.contrib.concurrent import process_map\n",
    "\n",
    "import fix_missing_residues\n",
    "\n",
    "pdb_structures = sorted(glob.glob('../data/pdb_downloaded/*.pdb') +\n",
    "                        glob.glob('../data/pdb_downloaded/*.cif'))\n",
    "\n",
    "def check_pdb(structure):\n",
    "    pdb_id = os.path.splitext(os.path.basename(structure))[0]\n",
    "    missing_residues = fix_missing_residues.check_missing_residues(structure)\n",
    "    return pdb_id, *missing_residues\n",
    "\n",
    "\n",
    "pdb_missing_residues = process_map(check_pdb, pdb_structures, chunksize=1)\n",
    "\n",
    "df_missing_residues = pd.DataFrame(pdb_missing_residues, columns=['PDB ID', 'Has Missing Residues', 'Missing Residues'])\n",
    "df_missing_residues.value_counts('Has Missing Residues')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e828a41-5e97-4015-a5f7-8c7028e34812",
   "metadata": {},
   "source": [
    "## Find PDB Structures with Multiple Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "195f86d1-e58d-452f-9053-719754281feb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_models_and_chains(pdb_file):\n",
    "    pdb_id = os.path.splitext(os.path.basename(pdb_file))[0]\n",
    "    structure = fix_missing_residues.parse_structure(pdb_file)\n",
    "    # for chain in structure.get_chains():\n",
    "    num_chains = len(list(structure.get_chains()))\n",
    "    num_models = len(list(structure.get_models()))\n",
    "    return pdb_id, num_models, num_chains\n",
    "\n",
    "model_chain_count = process_map(count_models_and_chains, pdb_structures, chunksize=1)\n",
    "df_num_models_chains = pd.DataFrame(model_chain_count, columns=['PDB ID', 'Number of Models', 'Number of Chains'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84849abb-d54b-4ff3-8a97-136a28c42110",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_num_models_chains.value_counts(['Number of Chains']).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c672d2-f4a4-4ba1-a59c-410f1e53cbdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_num_models[df_num_models['Number of Chains'] > 25]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ec3ce1a-8a31-4ac5-abff-9ed1bd7362fc",
   "metadata": {},
   "source": [
    "## Get Catalytic Site Atlas Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f6fdc9-0c91-469f-89b2-a20ca3cd7dcc",
   "metadata": {},
   "source": [
    "1. Get the manually curated catalytic residues from the Mechanism and Catalytic Site Atlas (M-CSA)\n",
    "2. Parse the JSON file and create a pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5e1b2cc-f167-4d63-88ff-0697f8c34e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get('https://www.ebi.ac.uk/thornton-srv/m-csa/api/residues/?format=json')\n",
    "mcsa_data = response.json()\n",
    "\n",
    "output = []\n",
    "for i, entry in enumerate(mcsa_data):\n",
    "    if entry['residue_chains']:\n",
    "        chain_name          = entry['residue_chains'][0]['chain_name']\n",
    "        pdb_id              = entry['residue_chains'][0]['pdb_id'].upper()\n",
    "        assembly_chain_name = entry['residue_chains'][0]['assembly_chain_name']\n",
    "        assembly            = entry['residue_chains'][0]['assembly']\n",
    "        code                = entry['residue_chains'][0]['code']\n",
    "        resid               = entry['residue_chains'][0]['resid']\n",
    "    else:\n",
    "        chain_name          = ''\n",
    "        pdb_id              = ''\n",
    "        assembly_chain_name = ''\n",
    "        assembly            = ''\n",
    "        code                = ''\n",
    "        resid               = float('nan')\n",
    "    uniprot_id          = entry['residue_sequences'][0]['uniprot_id']\n",
    "    res_seq_code        = entry['residue_sequences'][0]['code']\n",
    "    res_seq_resid       = entry['residue_sequences'][0]['resid']\n",
    "\n",
    "    output.append([chain_name, pdb_id, assembly_chain_name, assembly,\n",
    "                   code, resid, uniprot_id, res_seq_code, res_seq_resid])\n",
    "\n",
    "df_csa = pd.DataFrame(output, columns=['chain_name', 'pdb_id',\n",
    "                                       'assembly_chain_name', 'assembly',\n",
    "                                       'code', 'resid', 'uniprot_id',\n",
    "                                       'res_seq_code', 'res_seq_resid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5edb79aa-bce3-4256-9071-a085eb765514",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the UniProt and PDB IDs uppercase to facilitate merging on these later\n",
    "df_csa['uniprot_id'] = df_csa['uniprot_id'].str.upper()\n",
    "df_csa['pdb_id'] = df_csa['pdb_id'].str.upper()\n",
    "\n",
    "# Replace the assembly_chain_name with chain_name if there are odd chain_name in assembly_chain_name column\n",
    "df_csa.loc[df_csa['assembly_chain_name'].str.len() > 1, 'assembly_chain_name'] = df_csa['chain_name']\n",
    "print('Number of UniProt IDs in M-CSA:', df_csa['uniprot_id'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c9d5cf8-9ecd-4696-8a1c-be32c28ad197",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of Rows:             ', df_csa.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_csa['pdb_id'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_csa['uniprot_id'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faa58820-65e9-4d07-8982-ee00aab23a23",
   "metadata": {},
   "source": [
    "3. Download M-CSA PDB IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d00c6866-62c1-4ec4-bf7b-b50a98e1d439",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniprot_from_pdb_chains(pdb_chains):\n",
    "    pdb_chains_string = '\", \"'.join(pdb_chains)\n",
    "    \n",
    "    body = 'query {polymer_entity_instances(instance_ids: [\"' + pdb_chains_string + '\"])' + \"\"\"\n",
    "      {\n",
    "        rcsb_id\n",
    "        polymer_entity {\n",
    "          uniprots {\n",
    "            rcsb_id\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \"\"\"\n",
    "    response = requests.post(url='https://data.rcsb.org/graphql', json={\"query\": body})\n",
    "    response_data = response.json()\n",
    "\n",
    "    id_mapping = []\n",
    "    for record in response_data['data']['polymer_entity_instances']:\n",
    "        rcsb_id, chain = record['rcsb_id'].split('.')\n",
    "        if record['polymer_entity']['uniprots'] is not None:\n",
    "            uniprot_id = record['polymer_entity']['uniprots'][0]['rcsb_id']\n",
    "        else:\n",
    "            uniprot_id = ''\n",
    "        id_mapping.append([rcsb_id, chain, uniprot_id])\n",
    "        \n",
    "    return pd.DataFrame(id_mapping, columns=['pdb_id', 'chain', 'uniprot_id'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a74c33b-3911-4732-ab8c-a015324da241",
   "metadata": {},
   "source": [
    "### Update obsolete UniProt IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61216d94-3aab-4f5d-a7ac-76f42466668b",
   "metadata": {},
   "outputs": [],
   "source": [
    "csa_pdb_chains = set(df_csa['pdb_id'] + '.' + df_csa['assembly_chain_name'])\n",
    "df_csa_pdb = uniprot_from_pdb_chains(csa_pdb_chains)\n",
    "df_csa_pdb = df_csa_pdb.rename(columns={'chain': 'assembly_chain_name', 'uniprot_id': 'UniProt_AC'})\n",
    "df_csa_merged = df_csa.merge(df_csa_pdb, how='left', on=['pdb_id', 'assembly_chain_name'])\n",
    "df_csa_merged.loc[(df_csa_merged['uniprot_id'] != df_csa_merged['UniProt_AC']) & ~df_csa_merged['UniProt_AC'].isna(), 'uniprot_id'] = df_csa_merged['UniProt_AC']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5866425e-4125-40a3-875a-662887d620a5",
   "metadata": {},
   "source": [
    "### Mismatched UniProt IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d622fc7-fbca-45df-9511-bb47a2dfdf1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csa_merged_subset = df_csa_merged[['chain_name', 'pdb_id', 'assembly_chain_name', 'uniprot_id', 'UniProt_AC']].drop_duplicates()\n",
    "with pd.option_context('display.max_rows', 100, 'display.max_columns', 10):\n",
    "    display(\n",
    "        df_csa_merged_subset[df_csa_merged_subset['uniprot_id'] != df_csa_merged_subset['UniProt_AC']]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd8c196-c30c-4d9f-ae08-3109374146d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csa_pdb_uniprot = df_csa_merged_subset[['pdb_id', 'uniprot_id']].drop_duplicates()\n",
    "print('Number of Rows:             ', df_csa_pdb_uniprot.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_csa_pdb_uniprot['pdb_id'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_csa_pdb_uniprot['uniprot_id'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed0bd94-408f-4e88-9549-ae5337135b0b",
   "metadata": {},
   "source": [
    "3. Filter the M-CSA data for proteins contained in the ASD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c74257-c321-42b0-ab8d-2938ed05ce7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csa_in_asd = df_csa_merged[df_csa_merged['UniProt_AC'].isin(df_asd_updated['pdb_uniprot'])].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c012045e-8b74-4e5f-b30f-40b1fb358ae0",
   "metadata": {},
   "source": [
    "4. Group by PDB ID and UniProt ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd85ac00-63fd-4e29-9dca-4a405b29d37d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csa_in_asd['catalytic_site'] = (df_csa_in_asd['assembly_chain_name'] + '-'\n",
    "                                   + df_csa_in_asd['res_seq_resid'].astype(str)\n",
    "                                   + '-' + df_csa_in_asd['res_seq_code']\n",
    "                                   )\n",
    "\n",
    "df_csa_in_asd['catalytic_site_resids'] = (\n",
    "    df_csa_in_asd['res_seq_resid'])\n",
    "\n",
    "df_csa_in_asd_grouped = df_csa_in_asd[\n",
    "    ['UniProt_AC', 'catalytic_site', 'catalytic_site_resids']\n",
    "].groupby(['UniProt_AC']).agg(set).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5cd6c3-d30f-49d4-8474-28c5c2e4705a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csa_in_asd_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca0865ec-1ee3-4861-9865-659c312e632f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of Unique UniProt AC:', df_csa_in_asd_grouped['UniProt_AC'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "388b3427-95d7-459e-ac67-7460c784f63c",
   "metadata": {},
   "source": [
    "## Download the Active and Binding Site Data from UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3926212-ede1-4391-a786-97ad3628a1ff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import itertools\n",
    "import tqdm\n",
    "import SPARQLWrapper # import JSON, SPARQLWrapper\n",
    "\n",
    "import sys\n",
    "\n",
    "if sys.version_info >= (3, 12):\n",
    "    from itertools import batched\n",
    "else:\n",
    "    try:\n",
    "        from more_itertools import batched\n",
    "    except ImportError:\n",
    "        def batched(iterable, chunk_size):\n",
    "            iterator = iter(iterable)\n",
    "            while chunk := tuple(itertools.islice(iterator, chunk_size)):\n",
    "                yield chunk\n",
    "\n",
    "\n",
    "def get_uniprot_site_annotations(uniprot_ids):\n",
    "    # Set the SPARQL endpoint (UniProt)\n",
    "    sparql = SPARQLWrapper.SPARQLWrapper(\"https://sparql.uniprot.org/sparql\")\n",
    "\n",
    "    output = []\n",
    "    for uniprot_subset in tqdm.tqdm(itertools.batched(uniprot_ids, 200)):\n",
    "        uniprot_string = ' '.join([f'uniprotkb:{id}' for id in uniprot_subset])\n",
    "                 \n",
    "        # Define the query\n",
    "        query_string = f\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX faldo: <http://biohackathon.org/resource/faldo#>\n",
    "SELECT DISTINCT ?uniprot_id ?begin ?end ?site ?comment\n",
    "WHERE\n",
    "{{\n",
    "    VALUES ?protein {{ {uniprot_string} }}\n",
    "    BIND(substr(str(?protein), strlen(str(uniprotkb:))+1) AS ?uniprot_id)\n",
    "  \n",
    "    ?protein up:annotation ?annotation .\n",
    "  {{ ?annotation a up:Binding_Site_Annotation }} UNION {{ ?annotation a up:Active_Site_Annotation }} .\n",
    "    ?annotation rdf:type ?type .\n",
    "    BIND(substr(str(?type), strlen(str(up:))+1) AS ?site)\n",
    "    ?annotation up:range ?range .\n",
    "    ?range faldo:begin/faldo:position ?begin .\n",
    "    ?range faldo:end/faldo:position ?end .\n",
    "    OPTIONAL\n",
    "    {{\n",
    "        ?annotation up:ligand ?ligand .\n",
    "        ?ligand rdfs:comment ?comment .\n",
    "    }}\n",
    "}}\n",
    "\"\"\"\n",
    "        sparql.setQuery(query_string)\n",
    "    \n",
    "        # Set the output format as JSON\n",
    "        sparql.setReturnFormat(SPARQLWrapper.JSON)\n",
    "        \n",
    "        # Run the SPARQL query and convert to the defined format\n",
    "        data = sparql.query().convert()\n",
    "\n",
    "        # Store the query result\n",
    "        for result in data[\"results\"][\"bindings\"]:\n",
    "            output.append({key: value['value'] for key, value in result.items()})\n",
    "    return pd.DataFrame(output, columns=['uniprot_id', 'site', 'begin', 'end', 'comment'])\n",
    "\n",
    "\n",
    "def get_uniprot_sequence(uniprot_ids):\n",
    "    # Set the SPARQL endpoint (UniProt)\n",
    "    sparql = SPARQLWrapper.SPARQLWrapper(\"https://sparql.uniprot.org/sparql\")\n",
    "\n",
    "    output = []\n",
    "    for uniprot_subset in tqdm.tqdm(itertools.batched(uniprot_ids, 200)):\n",
    "        uniprot_string = ' '.join([f'uniprotkb:{id}' for id in uniprot_subset])\n",
    "                 \n",
    "        # Define the query\n",
    "        query_string = f\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT DISTINCT ?uniprot_id ?sequence\n",
    "WHERE\n",
    "{{\n",
    "    VALUES ?protein {{ {uniprot_string} }}\n",
    "    BIND(substr(str(?protein), strlen(str(uniprotkb:))+1) AS ?uniprot_id)\n",
    "    ?protein up:sequence ?isoform .\n",
    "    ?isoform a up:Simple_Sequence ;\n",
    "    rdf:value ?sequence .\n",
    "}}\n",
    "\"\"\"\n",
    "        sparql.setQuery(query_string)\n",
    "    \n",
    "        # Set the output format as JSON\n",
    "        sparql.setReturnFormat(SPARQLWrapper.JSON)\n",
    "        \n",
    "        # Run the SPARQL query and convert to the defined format\n",
    "        data = sparql.query().convert()\n",
    "\n",
    "        # Store the query result\n",
    "        for result in data[\"results\"][\"bindings\"]:\n",
    "            output.append({key: value['value'] for key, value in result.items()})\n",
    "    return pd.DataFrame(output, columns=['uniprot_id', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d682d8e-afd8-41c2-8b48-aebbc78a6803",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_ids = {uniprot_id for uniprot_id_list in df_asd_pdb_merged['uniprot_in_pdb'].drop_duplicates() for uniprot_id in uniprot_id_list}\n",
    "\n",
    "df_uniprot_site_annotations = get_uniprot_site_annotations(uniprot_ids)\n",
    "df_uniprot_sequences = get_uniprot_sequence(uniprot_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b89ea5c-d655-470b-b5cc-5bbc7147d3b5",
   "metadata": {},
   "source": [
    "Fix multiple simple sequences returned by UniProt SparQL and use UniProt REST API to fetch the canonical fasta sequence and remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f0e785-ebb1-46fa-a385-8c40607ef5df",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of Unique UniProt AC:', df_uniprot_site_annotations['uniprot_id'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899334f6-fb08-490b-b61d-3fea47519310",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib_venn\n",
    "\n",
    "matplotlib_venn.venn2([set(df_csa_in_asd_grouped['UniProt_AC']),\n",
    "                       set(df_uniprot_site_annotations['uniprot_id'])],\n",
    "                      ['M-CSA', 'UniProt']\n",
    "                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1725857-027e-4947-b6aa-c577b3d09e2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76e173d6-2745-4145-b333-74eb8222b48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_duplicated_sequences = df_uniprot_sequences[df_uniprot_sequences.duplicated('uniprot_id', keep=False)]\n",
    "for uniprot_id in df_duplicated_sequences['uniprot_id'].unique():\n",
    "    response = requests.get(f'https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta')\n",
    "    sequence = ''.join(response.text.split('\\n')[1:])\n",
    "    df_uniprot_sequences.loc[df_uniprot_sequences['uniprot_id'] == uniprot_id, 'sequence'] = sequence\n",
    "df_uniprot_sequences.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "484437c0-a04b-41ec-bd89-cde72004b5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniprot_site_annotations.fillna('', inplace=True)\n",
    "df_uniprot_site_annotations[df_uniprot_site_annotations['comment'].str.contains('alloster')].value_counts('comment')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda5b8fd-c448-44c9-a1db-ae02296521d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of unique UniProt Accesstion Numbers\n",
    "df_uniprot_site_annotations.loc[df_uniprot_site_annotations['comment'].str.contains('alloster'), 'uniprot_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12bfe339-2add-473b-9d79-7c8701094b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniprot_site_annotations[['comment']] = df_uniprot_site_annotations[['comment']].fillna('')\n",
    "\n",
    "df_active_site_annotations = df_uniprot_site_annotations[~df_uniprot_site_annotations['comment'].str.contains('alloster')].copy()\n",
    "\n",
    "df_active_site_annotations['begin'] = df_active_site_annotations['begin'].astype(int)\n",
    "df_active_site_annotations['end'] = df_active_site_annotations['end'].astype(int)\n",
    "\n",
    "# Group the annotation of each UniProt ID in multiple rows \n",
    "active_sites = []\n",
    "for index, group in df_active_site_annotations.groupby(['uniprot_id']):\n",
    "    active_site_residues = []\n",
    "    for _, record in group.iterrows():\n",
    "        # Expand the annotation ranges to list of residue indices\n",
    "        residues = list(range(record['begin'], record['end'] + 1))\n",
    "        active_site_residues.extend(residues)\n",
    "    active_sites.append([index[0], active_site_residues])\n",
    "\n",
    "df_active_sites = pd.DataFrame(active_sites, columns=['uniprot_id', 'active_site_residues'])\n",
    "\n",
    "df_active_sites = df_uniprot_sequences.merge(df_active_sites, how='left')\n",
    "df_active_sites[['active_site_residues']] = df_active_sites[['active_site_residues']].fillna('[]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddceb133-9887-4f74-b9f0-aa663d1bc1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_active_sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d2ee770-7b59-4e12-95e4-6e1cdb983d02",
   "metadata": {},
   "source": [
    "## Combine M-CSA and UniProt\n",
    "\n",
    "**Note:** Residue indices listed in M-CSA are in agreement with UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da459261-338a-4ed6-83b7-4d32fc315ec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_active_sites['active_site_residues'] = df_active_sites['active_site_residues'].apply(eval)\n",
    "df_active_sites.rename(columns={'uniprot_id': 'UniProt_AC'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7358260-ba5c-4939-8fbb-ebe1dd9a5819",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniprot_csa = df_uniprot.merge(df_csa_in_asd_grouped, how='left', on='UniProt_AC')\n",
    "\n",
    "active_site_residues = []\n",
    "\n",
    "for ind, record in df_uniprot_csa.iterrows():\n",
    "    if record['catalytic_site_resids'] is not float('nan'):\n",
    "        catalytic_site_resids = record['catalytic_site_resids']\n",
    "    else:\n",
    "        catalytic_site_resids = set()\n",
    "    print(type(record['active_site_residues']), type(catalytic_site_resids))\n",
    "    active_site_residues.append(sorted(set(record['active_site_residues']) | catalytic_site_resids))\n",
    "\n",
    "df_uniprot_csa['active_site_residues'] = active_site_residues\n",
    "\n",
    "# Remove rows with no active site residues\n",
    "df_uniprot_csa = df_uniprot_csa[(df_uniprot_csa['active_site_residues'].astype(str) != '[]')]\n",
    "\n",
    "df_uniprot_csa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2cf4186-e01b-4a56-b4cb-9475a2a87862",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_asd_pdb_merged['uniprot_in_pdb'].apply(len).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3d218f-6d45-4b13-8d1e-52bdb917380b",
   "metadata": {},
   "source": [
    "## Copy the allosteric site annotation to identical chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00bb9864-adcc-4b3d-8cb8-6491f5715bfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "import itertools\n",
    "\n",
    "allosteric_sites = []\n",
    "for index, record in tqdm(df_asd_with_active_site.iterrows(), total=df_asd_with_active_site.shape[0]):\n",
    "    protein_allosteric_sites = []\n",
    "\n",
    "    site_residues = []\n",
    "    for residue in record['allosteric_site_residues']:\n",
    "        residue = residue.upper()  # Change any lower case residue names or chains to upper case\n",
    "        chain = residue[0]\n",
    "        # Convert chains like '1', '2', ... to 'A', 'B', ...\n",
    "        if chain.isnumeric():\n",
    "            chain = chr(64 + int(chain))\n",
    "            residue = chain + residue[1:]  # Update the chain in the residue\n",
    "        site_residues.append(residue)\n",
    "\n",
    "    # Create a dictionary with chains with each uniprot id\n",
    "    uniprot_to_chains = collections.defaultdict(list)\n",
    "    chains_to_uniprot = collections.defaultdict(list)\n",
    "    for chain, uniprots, seq in record['chain_uniprot_mapping']:\n",
    "        for uniprot_id in uniprots:\n",
    "            uniprot_to_chains[uniprot_id].append(chain)\n",
    "            chains_to_uniprot[chain].append(uniprot_id)\n",
    "\n",
    "    for chain, residues in itertools.groupby(site_residues, key=lambda res: res[0]):\n",
    "        residue_list = list(residues)\n",
    "        for uniprot in chains_to_uniprot[chain]:\n",
    "            for identical_chain in uniprot_to_chains[uniprot]:\n",
    "                protein_allosteric_sites.append([identical_chain, [identical_chain + res[1:] for res in residue_list]])\n",
    "    allosteric_sites.append(protein_allosteric_sites)\n",
    "\n",
    "df_asd_with_active_site['allosteric_sites'] = allosteric_sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5a9c9af-746b-4698-9e5e-0c05a763f129",
   "metadata": {},
   "source": [
    "## Add active site annotation to identical chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dadb6be-8bae-48df-95e1-e95562ae0e1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "active_sites = []\n",
    "for index, record in tqdm(df_asd_with_active_site.iterrows(), total=df_asd_with_active_site.shape[0]):\n",
    "    protein_active_sites = []\n",
    "    for chain in record['chain_uniprot_mapping']:\n",
    "        for uniprot_id in chain[1]:\n",
    "            active_site = df_uniprot_csa.loc[df_uniprot_csa['UniProt_AC'] == uniprot_id, 'active_site_residues']\n",
    "            if not active_site.empty:\n",
    "                protein_active_sites.append([chain[0], list(active_site)[0]])\n",
    "    active_sites.append(protein_active_sites)\n",
    "\n",
    "df_asd_with_active_site['active_sites'] = active_sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948604e3-9bfa-4af8-ba31-59b7ff5a5010",
   "metadata": {},
   "source": [
    "Filter the dataframe with at least one allosteric site reidue and one active or binding site residue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64adf9f-b7c9-4f1d-a625-5bb8f46463e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop the rows with no active site residues\n",
    "df_asd_with_active_site = df_asd_with_active_site[df_asd_with_active_site['active_sites'].str.len() != 0]\n",
    "\n",
    "# Drop the rows with no allosteric site residues\n",
    "df_asd_with_active_site = df_asd_with_active_site[df_asd_with_active_site['allosteric_sites'].str.len() != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b2345b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of Rows:             ', df_asd_with_active_site.shape[0])\n",
    "print('Number of Unique PDB IDs:   ', df_asd_with_active_site['allosteric_pdb'].nunique())\n",
    "print('Number of Unique UniProt AC:', df_asd_with_active_site['pdb_uniprot'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5483df7f-ac74-4112-84c9-212e4979364f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_asd_with_active_site.to_csv('../output/ASD_with_Active_Sites.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allostery",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
